streams_map_incoming_bidi.go

```diff
@@ -69,18 +69,20 @@ func (m *incomingBidiStreamsMap) AcceptStream() (streamI, error) {
 }
 
 func (m *incomingBidiStreamsMap) GetOrOpenStream(id protocol.StreamID) (streamI, error) {
+	m.mutex.RLock()
 	if id > m.maxStream {
+		m.mutex.RUnlock()
 		return nil, fmt.Errorf("peer tried to open stream %d (current limit: %d)", id, m.maxStream)
 	}
 	// if the id is smaller than the highest we accepted
 	// * this stream exists in the map, and we can return it, or
 	// * this stream was already closed, then we can return the nil
 	if id <= m.highestStream {
-		m.mutex.RLock()
 		s := m.streams[id]
 		m.mutex.RUnlock()
 		return s, nil
 	}
+	m.mutex.RUnlock()
```

Comment:
I think there's another race here. You need to do these checks again once you've acquired the write lock, as another thread might have just completed the write locked block.

Reply:
I don't think this is necessary here.
`m.maxStream` can only be increased (by `DeleteStream`), so the id is valid for the first check (`id <= m.maxStream`), it will definitely be valid if `m.maxStream` is increased.
Regarding `m.highestStream`, that variable is only read by other methods, never written, so the value can't change.

Reply:
Then add a comment stating that. That makes it clear that you thought about it, and ensures we don't forget about this assumption if we make later changes.

Reply:
Done.

Reply:
Does the same apply for the other two locations?

Reply:
The other two are auto generated from this one.

Reply:
Ah, right 🙄